src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,256 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines.Networking.Tls.Internal.OpenSsl
+{
+    internal static unsafe class CustomBio
+    {
+        //Max Pipeline block size at the moment (4k - 1 cache line for tracking info) 
+        //this should come from the library directly when available in the future
+        private const int MaxBlockSize = 1024 * 4 - 64;
+        private const int BIO_TYPE_MEM = 1 | 0x0400 | 2;
+        private const int BIO_TYPE_SOURCE_SINK = 0x0400;
+
+        private static readonly bio_method_st s_methodStruct;
+        private static readonly IntPtr s_methodPtr;
+        private static readonly Create s_create;
+        private static readonly Write s_write;
+        private static readonly Read s_read;
+        private static readonly Control s_control;
+        private static readonly Free s_free;
+
+        static CustomBio()
+        {
+            s_create = CreateBio;
+            s_write = WriteBio;
+            s_read = ReadBio;
+            s_control = ControlBio;
+            s_free = FreeBio;
+
+            s_methodStruct = new bio_method_st()
+            {
+                create = s_create,
+                breadDelegate = s_read,
+                bwriteDelegate = s_write,
+                ctrlDelegate = s_control,
+                type = BIO_TYPE_MEM,
+                destroy = s_free,
+                name = null
+            };
+            var sizeToAlloc = Marshal.SizeOf(s_methodStruct);
+            s_methodPtr = Marshal.AllocHGlobal(sizeToAlloc);
+            Marshal.StructureToPtr(s_methodStruct, s_methodPtr, false);
+        }
+
+        public static IntPtr Custom() => s_methodPtr;
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Create(ref bio_st bio);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Write(ref bio_st bio, void* buf, int num);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Read(ref bio_st bio, void* buf, int size);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate long Control(ref bio_st bio, BioControl cmd, long num, void* ptr);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Free(ref bio_st data);
+
+        private static class WindowsLib
+        {
+            public const string CryptoDll = "libeay32.dll";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
```

Comment:
The plan here I believe is to "copy" (because for now its not in corefx but labs) all of the shim code across from the current crypto and reuse that.